### DISCUSSION TITLE
Using a basestation facing upwards, and the crazyflie upside down.

### DISCUSSION THREAD

**Discussion by cafeciaojoe** [bitcraze/discussions#1518] [Q&A]
Hello all! 

I have conducted a basic test to see how stable the lighthouse system is when the are three basestations, with one of the base stations facing upwards. The test involves roughly estimating the variance of the "stablilizer.roll" and the "stateEstimateZ.z" variables whilst the crazyflie is mounted on a tripod facing up (about 0deg roll), sideways (about -90deg roll), and upside down ( about -135deg roll). 

The system was calibrated using the wizard, in the latest version of the client and using the latest firmware on a cf 2.1. The third basestation was not visible for the first three steps of the wizard (ie origin, origin + 1m and n number of XY plane measurements)

I encountered some major variance during this experiment when the crazyflie was using the basestation facing up. 

There was some insane "oscillating variance" (9deg roll and 200mm height), that was fixed with a restart of the crazyflie, but then some significant and slightly more random variance after the restart (not in the roll, but 50mm in the .z)

if you scrub thru this video you can see how I collected the data. Let me know how I can improve this experiment and provide better data for our diagnosis

https://youtu.be/GtufDYDOZto

<img width="703" alt="Screenshot 2024-08-16 at 17 45 10" src="https://github.com/user-attachments/assets/9b3d9230-845e-4a1b-bcb8-6e70bcc07f71">

Thanks all !
Joe

**Comment by gemenerik** [comment]
We didn’t design LH for this specific use case and we seem to remember making some assumptions regarding orientation so pretty cool to see you got an upward facing base station geometry estimate.

We don’t estimate roll and pitch using lighthouse data. It’s crucial to have accurate estimates for these parameters for LH positioning to function correctly. In the 90-degree BS3 example, it seems that everything works reasonably well until the Kalman filter resets the roll, which then causes everything to break down. This likely explains why the oscillations disappear after the restart. We’re still unsure about the cause of this.

It would be helpful if you could [log the data to a file](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/logblocks_tab/) so we can analyze what’s happening in more detail. In addition to roll and z estimates, it would be great to see the full orientation and position estimates. A video like the one you provided is also very useful for correlating the data with the real-world events.

**Comment by cafeciaojoe** [comment]
Great thanks for the reply! 

I have conducted the test again and exported the logs, I have also included the calibration file. 

Here is the video: https://youtu.be/0h4xMU1z_gU


[lhStatus-20240819T15-50-50.csv](https://github.com/user-attachments/files/16661203/lhStatus-20240819T15-50-50.csv)
[Pose-20240819T15-50-51.csv](https://github.com/user-attachments/files/16661204/Pose-20240819T15-50-51.csv)
[stabilizer-20240819T15-50-52.csv](https://github.com/user-attachments/files/16661205/stabilizer-20240819T15-50-52.csv)
[1 bs 2 downv1 up.yaml.zip](https://github.com/user-attachments/files/16661216/1.bs.2.downv1.up.yaml.zip)

**Reply by gemenerik** [reply]
Thanks for the additional data! We're currently tied up with next quarter's planning, so it might take a bit of time to address this issue fully. We’ll get back to you when we can. In the meantime, you might want to start investigating the lead about the Kalman filter reset.

**Comment by cafeciaojoe** [comment]
Hello all :)

I have spend some time over the last few days on this problem. I have broken it down into two parts. 

### Problem A

The calibration wizard fails I would say 4 out of 5 times when calibrating at the Bitcraze Lab. Compared to my lab. Marcus said that there are quite possibility too many reflections from the truss, lights, and the tv screen. 

If you want to start replicating the roll test from the previous post have left a little set up where perhaps we could run a calibration test under a table with a table cloth over, just to be sure. (lol) 

One other things that may be causing this to fail which I will test back at my lab.

- the crazyflie going past 90 degrees pitch forward when doing the final freeform calibration.

Here is a video and the log files logdata from the one successful calibration. 

https://youtu.be/ZAAM3yZfLJA

[__calibration__lhStatus-20240926T16-11-10.csv](https://github.com/user-attachments/files/17166166/__calibration__lhStatus-20240926T16-11-10.csv)
[__calibration__Pose-20240926T16-11-11.csv](https://github.com/user-attachments/files/17166167/__calibration__Pose-20240926T16-11-11.csv)
[__calibration__stabilizer-20240926T16-11-12.csv](https://github.com/user-attachments/files/17166168/__calibration__stabilizer-20240926T16-11-12.csv)


### Problem B

With the one successful calibration I did perform I also did a pitch test. Below is the video and the log data for that for that.

https://youtu.be/MsQ-RZgoQ1s

[__pitch_test__lhStatus-20240926T16-46-58.csv](https://github.com/user-attachments/files/17166210/__pitch_test__lhStatus-20240926T16-46-58.csv)
[__pitch_test__Pose-20240926T16-46-57.csv](https://github.com/user-attachments/files/17166211/__pitch_test__Pose-20240926T16-46-57.csv)
[__pitch_test__stabilizer-20240926T16-46-57.csv](https://github.com/user-attachments/files/17166212/__pitch_test__stabilizer-20240926T16-46-57.csv)

Taking a quick look at the data you can see that after I pick up the drone and walk in the positive x direction. Things get crazy and the data points are connected by straight lines, then I go and put the cf back at the origin. 

![Figure_3](https://github.com/user-attachments/assets/7802c7fc-0e8d-4a60-ab1f-af53ddb40436)

Interestingly the stateEstimate.roll basically flips at some point. I have not yet correlated this with the lighthouse.bsActive (I *think* that is the variable I need to be looking at)

![Figure_1](https://github.com/user-attachments/assets/8b642c2b-b6ae-491e-bf65-a27c54c93e62)
![Figure_2](https://github.com/user-attachments/assets/782c49ea-4082-4eb3-ad6b-173b1c860abd)

Thank you for your time!

**Comment by knmcguire** [comment]
This needs a larger investigation so we made an ticket for this, now in the firmware: https://github.com/bitcraze/crazyflie-firmware/issues/1413.

We will need to find time to work on this for a long duration of time since this is much more difficult that we initially anticipated. Hopefully the next 3rd week cycle.
